src/adapter/src/coord/sequencer/inner/cluster.rs

```diff
@@ -238,12 +282,119 @@ impl Coordinator {
+        let mut ops = vec![];
+
+        // Gather the ops to remove the non pending replicas
+        // Also skip any billed_as free replicas
+        let remove_replicas = cluster
+            .replicas()
+            .filter_map(|r| {
+                if !r.config.location.pending() && r.config.location.billed_as().is_none() {
+                    Some(catalog::DropObjectInfo::ClusterReplica((
+                        cluster_id.clone(),
+                        r.replica_id,
+                        ReplicaCreateDropReason::Manual,
+                    )))
+                } else {
+                    None
+                }
+            })
+            .collect();
+        ops.push(catalog::Op::DropObjects(remove_replicas));
+
+        // Gather the Ops to remove the "-pending" suffix from the name and set
+        // pending to false
+        let finalize_replicas: Vec<catalog::Op> = cluster
+            .replicas()
+            .filter_map(|r| {
+                if r.config.location.pending() {
+                    let cluster_ident = match Ident::new(cluster.name.clone()) {
+                        Ok(id) => id,
+                        Err(err) => return Some(Err(err)),
+                    };
+                    let replica_ident = match Ident::new(r.name.clone()) {
+                        Ok(id) => id,
+                        Err(err) => return Some(Err(err)),
+                    };
+                    Some(Ok((cluster_ident, replica_ident, r)))
+                } else {
+                    None
+                }
+            })
+            // Early collection is to handle errors from generating of the
+            // Idents
+            .collect::<Result<Vec<(Ident, Ident, &ClusterReplica)>, _>>()?
+            .into_iter()
+            .map(|(cluster_ident, replica_ident, replica)| {
+                let mut new_replica_config = replica.config.clone();
+                debug!("Promoting replica: {}", replica.name);
+                match new_replica_config.location {
+                    mz_controller::clusters::ReplicaLocation::Managed(ManagedReplicaLocation {
+                        ref mut pending,
+                        ..
+                    }) => {
+                        *pending = false;
+                    }
+                    _ => {}
+                }
+
+                vec![
+                    catalog::Op::RenameClusterReplica {
+                        cluster_id: cluster_id.clone(),
+                        replica_id: replica.replica_id.to_owned(),
+                        name: QualifiedReplica {
+                            cluster: cluster_ident,
+                            replica: replica_ident,
+                        },
+                        to_name: replica
+                            .name
+                            .strip_suffix(PENDING_REPLICA_SUFFIX)
+                            .expect("Pending replicas must have names ending with -pending")
```

Comment:
The comment hard-codes the pending replica suffix, which will be confusing should it change.